src/core/qgspointlocator.cpp

```diff
@@ -134,6 +134,103 @@ class QgsPointLocator_VisitorNearestVertex : public IVisitor
 };
 
 
+
+/**
+ * \ingroup core
+ * Helper class used when traversing the index looking for centroid - builds a list of matches.
+ * \note not available in Python bindings
+ * \since QGIS 3.12
+*/
+class QgsPointLocator_VisitorNearestCentroid : public IVisitor
+{
+  public:
+    QgsPointLocator_VisitorNearestCentroid( QgsPointLocator *pl, QgsPointLocator::Match &m, const QgsPointXY &srcPoint, QgsPointLocator::MatchFilter *filter = nullptr )
+      : mLocator( pl )
+      , mBest( m )
+      , mSrcPoint( srcPoint )
+      , mFilter( filter )
+    {}
+
+    void visitNode( const INode &n ) override { Q_UNUSED( n ); }
+    void visitData( std::vector<const IData *> &v ) override { Q_UNUSED( v ); }
+
+    void visitData( const IData &d ) override
+    {
+      QgsFeatureId id = d.getIdentifier();
+      QgsGeometry *geom = mLocator->mGeoms.value( id );
+
+      QgsPointXY pt = geom->centroid().asPoint();
+
+      QgsPointLocator::Match m( QgsPointLocator::Centroid, mLocator->mLayer, id, std::sqrt( mSrcPoint.sqrDist( pt ) ), pt, -1 );
+      // in range queries the filter may reject some matches
+      if ( mFilter && !mFilter->acceptMatch( m ) )
+        return;
+
+      if ( !mBest.isValid() || m.distance() < mBest.distance() )
+        mBest = m;
+
+    }
+
+  private:
+    QgsPointLocator *mLocator = nullptr;
+    QgsPointLocator::Match &mBest;
+    QgsPointXY mSrcPoint;
+    QgsPointLocator::MatchFilter *mFilter = nullptr;
+};
+
+////////////////////////////////////////////////////////////////////////////
+
+/**
+ * \ingroup core
+ * Helper class used when traversing the index looking for middle segment - builds a list of matches.
+ * \note not available in Python bindings
+ * \since QGIS 3.12
+*/
+class QgsPointLocator_VisitorNearestMiddle: public IVisitor
+{
+  public:
+    QgsPointLocator_VisitorNearestMiddle( QgsPointLocator *pl, QgsPointLocator::Match &m, const QgsPointXY &srcPoint, QgsPointLocator::MatchFilter *filter = nullptr )
+      : mLocator( pl )
+      , mBest( m )
+      , mSrcPoint( srcPoint )
+      , mFilter( filter )
+    {}
+
+    void visitNode( const INode &n ) override { Q_UNUSED( n ); }
+    void visitData( std::vector<const IData *> &v ) override { Q_UNUSED( v ); }
+
+    void visitData( const IData &d ) override
+    {
+      QgsFeatureId id = d.getIdentifier();
+      QgsGeometry *geom = mLocator->mGeoms.value( id );
+      QgsPointXY pt;
+      int afterVertex;
+      double sqrDist = geom->closestSegmentWithContext( mSrcPoint, pt, afterVertex, nullptr, POINT_LOC_EPSILON );
```

Comment:
Hmm -- the closest segment won't necessarily equate to the closest segment's midpoint, will it? (Think a close but very long segment vs a slightly more distant but very short segment)

Reply:
I don't think so, All my manuals tests worked well (tm)
But I'm interested in exploring it further if it comes to be confirmed.